Cannot invoke String.length() because str is null error with ParentIdQueryBuilder

On serverless we observed a couple of 500's with the message `Cannot invoke \"String.length()\" because \"str\" is null\`, caused by this line: https://github.com/elastic/elasticsearch/blob/21b0caaf13cf1d0dde5a85ef2a44917d3481d539/modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentIdQueryBuilder.java#L70

ES version: 8.13.0

I tried reproducing the error in a serverless environment but I didn't get past the syntax validation (`type` cannot be null).

Relevant stack trace:
```
"Failed to execute phase [query], all shards failed; shardFailures {[<redacted>][<redacted>][0]: org.elasticsearch.transport.SendRequestTransportException: [es-es-search-<redacted>][indices:data/read/search[phase/query]]
...
Caused by: java.lang.NullPointerException: Cannot invoke \"String.length()\" because \"str\" is null
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeString(StreamOutput.java:402)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeString(StreamOutput.java:389)
  at org.elasticsearch.join@8.13.0/org.elasticsearch.join.query.ParentIdQueryBuilder.doWriteTo(ParentIdQueryBuilder.java:70)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.index.query.AbstractQueryBuilder.writeTo(AbstractQueryBuilder.java:76)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:995)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.search.builder.SubSearchSourceBuilder.writeTo(SubSearchSourceBuilder.java:84)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeWriteable(StreamOutput.java:983)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeCollection(StreamOutput.java:1052)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeCollection(StreamOutput.java:1042)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.search.builder.SearchSourceBuilder.writeTo(SearchSourceBuilder.java:294)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.common.io.stream.StreamOutput.writeOptionalWriteable(StreamOutput.java:972)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.search.internal.ShardSearchRequest.innerWriteTo(ShardSearchRequest.java:361)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.search.internal.ShardSearchRequest.writeTo(ShardSearchRequest.java:349)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.transport.OutboundMessage.serialize(OutboundMessage.java:71)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.transport.OutboundHandler.sendMessage(OutboundHandler.java:186)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.transport.OutboundHandler.sendRequest(OutboundHandler.java:105)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.transport.TcpTransport$NodeChannels.sendRequest(TcpTransport.java:322)
  at org.elasticsearch.server@8.13.0/org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:953)
  ... 178 more
```

--- a/docs/changelog/105693.yaml
+++ b/docs/changelog/105693.yaml
@@ -1,0 +1,6 @@
+pr: 105693
+summary: Fix error 500 on invalid `ParentIdQuery`
+area: Search
+type: bug
+issues:
+ - 105366

--- a/modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentIdQueryBuilder.java
+++ b/modules/parent-join/src/main/java/org/elasticsearch/join/query/ParentIdQueryBuilder.java
@@ -51,8 +51,8 @@ public final class ParentIdQueryBuilder extends AbstractQueryBuilder<ParentIdQue
     private boolean ignoreUnmapped = DEFAULT_IGNORE_UNMAPPED;
 
     public ParentIdQueryBuilder(String type, String id) {
-        this.type = type;
-        this.id = id;
+        this.type = requireValue(type, "[" + NAME + "] requires '" + TYPE_FIELD.getPreferredName() + "' field");
+        this.id = requireValue(id, "[" + NAME + "] requires '" + ID_FIELD.getPreferredName() + "' field");
     }
 
     /**

--- a/modules/parent-join/src/test/java/org/elasticsearch/join/query/ParentIdQueryBuilderTests.java
+++ b/modules/parent-join/src/test/java/org/elasticsearch/join/query/ParentIdQueryBuilderTests.java
@@ -156,6 +156,11 @@ public class ParentIdQueryBuilderTests extends AbstractQueryTestCase<ParentIdQue
         assertThat(e.getMessage(), containsString("[" + ParentIdQueryBuilder.NAME + "] no relation found for child [unmapped]"));
     }
 
+    public void testThrowsOnNullTypeOrId() {
+        expectThrows(IllegalArgumentException.class, () -> new ParentIdQueryBuilder(null, randomAlphaOfLength(5)));
+        expectThrows(IllegalArgumentException.class, () -> new ParentIdQueryBuilder(randomAlphaOfLength(5), null));
+    }
+
     public void testDisallowExpensiveQueries() {
         SearchExecutionContext searchExecutionContext = mock(SearchExecutionContext.class);
         when(searchExecutionContext.allowExpensiveQueries()).thenReturn(false);

